eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentPacket.scala

```diff
@@ -213,9 +213,12 @@ object OutgoingPaymentPacket {
    *         - the trampoline onion to include in final payload of a normal onion
    */
   def buildTrampolineToLegacyPacket(invoice: PaymentRequest, hops: Seq[NodeHop], finalPayload: PaymentOnion.FinalPayload): (MilliSatoshi, CltvExpiry, Sphinx.PacketAndSecrets) = {
-    val (firstAmount, firstExpiry, payloads) = hops.drop(1).reverse.foldLeft((finalPayload.amount, finalPayload.expiry, Seq[PaymentOnion.PerHopPayload](finalPayload))) {
+    // NB: the final payload will never reach the recipient, since the next-to-last node in the trampoline route will convert that to a legacy payment.
+    // We use the smallest final payload possible, otherwise we may overflow the trampoline onion size.
+    val dummyFinalPayload = PaymentOnion.createSinglePartPayload(finalPayload.amount, finalPayload.expiry, finalPayload.paymentSecret, None)
```

Comment:
I'm lost here. Why did you have to do this in a seemingly completely unrelated commit ("Include payment metadata in all invoices")? Did you actually make the onion overflow during tests? Why can't the next-to-last node pass the `metadata` when it builds the legacy payload. What does `legacy` mean here: non-trampoline or non-tlv? 😵‍💫 

```suggestion
    // NB: the final payload will never reach the recipient, since the next-to-last node in the trampoline route will convert that to a legacy payment.
    // We use the smallest final payload possible, otherwise we may overflow the trampoline onion size.
    val dummyFinalPayload = PaymentOnion.createSinglePartPayload(finalPayload.amount, finalPayload.expiry, finalPayload.paymentSecret, None)
```

Reply:
> Did you actually make the onion overflow during tests?

Yes that's exactly it, existing tests start failing because of the addition of the `payment_metadata` that ended up duplicated in the trampoline onion :/

> Why can't the next-to-last node pass the metadata when it builds the legacy payload

That's what it does, so we don't need to have the `payment_metadata` in the _trampoline_ payload for the last node (since that payload is completely ignored because we instead convert the payment to a non-trampoline one).

Here legacy means non-trampoline, I'll clarify that.

Reply:
I updated the comment in https://github.com/ACINQ/eclair/pull/2063/commits/87aca520f7da877e5d58db728c1a7c64cb27c1c4